File: solution/c#/Day04/Day04.Tests/ArticleTests.cs
using FluentAssertions;
using Xunit;

namespace Day04.Tests
{
    public class ArticleTests
    {
        private const string Author = "Pablo Escobar";
        private const string CommentText = "Amazing article !!!";

        private readonly Article _article = new(
            "Lorem Ipsum",
            "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore"
        );

        [Fact]
        public void Should_Add_Comment_In_An_Article()
        {
            _article.AddComment(CommentText, Author);

            _article.Comments
                .Should().HaveCount(1)
                .And.ContainSingle(comment => comment.Text == CommentText && comment.Author == Author);
        }

        [Fact]
        public void Should_Add_Comment_In_An_Article_Containing_Already_A_Comment()
        {
            const string newComment = "Finibus Bonorum et Malorum";
            const string newAuthor = "Al Capone";

            _article.AddComment(CommentText, Author);
            _article.AddComment(newComment, newAuthor);

            _article.Comments.Should().HaveCount(2);

            var lastComment = _article.Comments.Last();
            lastComment.Text.Should().Be(newComment);
            lastComment.Author.Should().Be(newAuthor);
        }

        public class Fail : ArticleTests
        {
            [Fact]
            public void When_Adding_An_Existing_Comment()
            {
                _article.AddComment(CommentText, Author);

                var act = () => _article.AddComment(CommentText, Author);
                act.Should().Throw<CommentAlreadyExistException>();
            }
        }
    }
}

File: solution/java/day04/src/test/java/blog/ArticleTests.java
package blog;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class ArticleTests {
    public static final String AUTHOR = "Pablo Escobar";
    private static final String COMMENT_TEXT = "Amazing article !!!";
    private Article article;

    @BeforeEach
    void setup() {
        article = new Article(
                "Lorem Ipsum",
                "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore"
        );
    }

    @Test
    void should_add_comment_in_an_article() throws CommentAlreadyExistException {
        article.addComment(COMMENT_TEXT, AUTHOR);

        assertThat(article.getComments()).hasSize(1);

        var comment = article.getComments().get(0);
        assertThat(comment.text()).isEqualTo(COMMENT_TEXT);
        assertThat(comment.author()).isEqualTo(AUTHOR);
    }

    @Test
    void should_add_comment_in_an_article_containing_already_a_comment() throws CommentAlreadyExistException {
        var newComment = "Finibus Bonorum et Malorum";
        var newAuthor = "Al Capone";

        article.addComment(COMMENT_TEXT, AUTHOR);
        article.addComment(newComment, newAuthor);

        assertThat(article.getComments()).hasSize(2);

        var lastComment = article.getComments().getLast();
        assertThat(lastComment.text()).isEqualTo(newComment);
        assertThat(lastComment.author()).isEqualTo(newAuthor);
    }

    @Nested
    class Fail {
        @Test
        void when__adding_an_existing_comment() throws CommentAlreadyExistException {
            article.addComment(COMMENT_TEXT, AUTHOR);

            assertThatThrownBy(() -> {
                article.addComment(COMMENT_TEXT, AUTHOR);
            }).isInstanceOf(CommentAlreadyExistException.class);
        }
    }
}
